src/device_trezor/trezor/protocol.cpp

```diff
@@ -0,0 +1,896 @@
+        for (const auto &it : rsig_data.rsig_parts()) {
+          rsig_buff += it;
+        }
+      }
+    }
+
+    if (!cn_deserialize(ack->tx_out(), tx_out)){
+      throw exc::ProtocolException("Cannot deserialize vout[i]");
+    }
+
+    if (!cn_deserialize(ack->out_pk(), out_pk)){
+      throw exc::ProtocolException("Cannot deserialize out_pk");
+    }
+
+    if (!cn_deserialize(ack->ecdh_info(), ecdh)){
+      throw exc::ProtocolException("Cannot deserialize ecdhtuple");
+    }
+
+    if (has_rsig && !is_req_bulletproof() && !cn_deserialize(rsig_buff, range_sig)){
+      throw exc::ProtocolException("Cannot deserialize rangesig");
+    }
+
+    if (has_rsig && is_req_bulletproof() && !cn_deserialize(rsig_buff, bproof)){
+      throw exc::ProtocolException("Cannot deserialize bulletproof rangesig");
+    }
+
+    m_ct.tx.vout.emplace_back(tx_out);
+    m_ct.tx_out_hmacs.push_back(ack->vouti_hmac());
+    m_ct.tx_out_pk.emplace_back(out_pk);
+    m_ct.tx_out_ecdh.emplace_back(ecdh);
+
+    if (!has_rsig){
+      return;
+    }
+
+    if (is_req_bulletproof()){
+      CHECK_AND_ASSERT_THROW_MES(m_ct.cur_batch_idx < m_ct.grouping_vct.size(), "Invalid batch index");
+      auto batch_size = m_ct.grouping_vct[m_ct.cur_batch_idx];
+      for (size_t i = 0; i < batch_size; ++i){
+        const size_t bidx = 1 + m_ct.cur_output_idx - batch_size + i;
+        CHECK_AND_ASSERT_THROW_MES(bidx < m_ct.tx_out_pk.size(), "Invalid out index");
+
+        rct::key commitment = m_ct.tx_out_pk[bidx].mask;
+        commitment = rct::scalarmultKey(commitment, rct::INV_EIGHT);
+        bproof.V.push_back(commitment);
+      }
+
+      m_ct.tx_out_rsigs.emplace_back(bproof);
+      if (!rct::bulletproof_VERIFY(boost::get<rct::Bulletproof>(m_ct.tx_out_rsigs.back()))) {
+        throw exc::ProtocolException("Returned range signature is invalid");
+      }
+
+    } else {
+      m_ct.tx_out_rsigs.emplace_back(range_sig);
+
+      if (!rct::verRange(out_pk.mask, boost::get<rct::rangeSig>(m_ct.tx_out_rsigs.back()))) {
+        throw exc::ProtocolException("Returned range signature is invalid");
+      }
+    }
+
+    m_ct.cur_batch_idx += 1;
+    m_ct.cur_output_in_batch_idx = 0;
+  }
+
+  std::shared_ptr<messages::monero::MoneroTransactionAllOutSetRequest> Signer::step_all_outs_set(){
+    return std::make_shared<messages::monero::MoneroTransactionAllOutSetRequest>();
+  }
+
+  void Signer::step_all_outs_set_ack(std::shared_ptr<const messages::monero::MoneroTransactionAllOutSetAck> ack, hw::device &hwdev){
+    m_ct.rv = std::make_shared<rct::rctSig>();
+    m_ct.rv->txnFee = ack->rv().txn_fee();
+    m_ct.rv->type = static_cast<uint8_t>(ack->rv().rv_type());
+    string_to_key(m_ct.rv->message, ack->rv().message());
+
+    // Extra copy
+    m_ct.tx.extra.clear();
+    auto extra = ack->extra();
+    auto extra_data = extra.data();
+    for(size_t i = 0; i < extra.size(); ++i){
+      m_ct.tx.extra.push_back(static_cast<uint8_t>(extra_data[i]));
+    }
```

Comment:
There's probably easier to avoid several reallocs, but very minor if you don't feel like changing it.

Reply:
I've added reserve() which should fix the reallocations.